[hot_reload] Give more info about why the runtime can't apply updates

lambdageek opened the pull request:
Help with investigations like https://github.com/dotnet/aspnetcore/issues/47836

lambdageek committed:
[hot_reload] Give more info about why the runtime can't apply updates

thaystg commented on src/mono/mono/component/hot_reload.c:
```diff
@@ -2510,7 +2510,14 @@ void
+		} else if (!mono_assembly_is_jit_optimizer_disabled (image_base->assembly)) {
+			msg = "assembly does not have a System.Diagnostics.DebuggableAttribute with the DebuggingModes.DisableOptimizations flag (editing Release build assemblies is not supported)";
+		}
+		mono_error_set_invalid_operation (error, "The assembly '%s' can not be edited or changed, because %s", image_base->name, msg);
```
If it doesn't enter in the if nor in the else if the message printed will be something like:
"The assembly 'sample.dll' can not be edited or changed, because "
Maybe this is weird?

szalapski replied:
Yeah, I would think you should have an else condition of some kind, like "of an unforeseen reason" or something similar.

thaystg commented:
I'm approving it despite my comment, because I'm not sure if you want to merge it for P6.

tmat commented on src/mono/mono/component/hot_reload.c:
```diff
@@ -2510,7 +2510,14 @@ void
+		int modifiable = 0;
+		const char *msg = "";
+		if (!hot_reload_update_enabled (&modifiable) || modifiable != MONO_MODIFIABLE_ASSM_DEBUG) {
+			msg = "environment variable DOTNET_MODIFIABLE_ASSEMBLIES=\"Debug\" is not set";
```
Can this include the current value of the env var?

lambdageek committed:
rework hot_reload_update_enabled

Add a version that always checks the environment variable and returns
unexpected values back to the caller.

Change the assembly_update_supported function to take a MonoError and
set it to a detailed error message directly.

lambdageek commented:
Refactored the code a little bit.  Now we can show unexpected `DOTNET_MODIFIABLE_ASSEMBLIES` values; also set a `MonoError` directly in `assembly_update_supported` function so that we are always in sync between the checks and the error messages.
